lib/pyinotify.py

```diff
@@ -0,0 +1,2346 @@
+                        'SPECIAL_FLAGS': {
+        'IN_ONLYDIR'       : 0x01000000,  # only watch the path if it is a
+                                          # directory
+        'IN_DONT_FOLLOW'   : 0x02000000,  # don't follow a symlink
+        'IN_EXCL_UNLINK'   : 0x04000000,  # exclude events on unlinked objects
+        'IN_MASK_ADD'      : 0x20000000,  # add to the mask of an already
+                                          # existing watch
+        'IN_ISDIR'         : 0x40000000,  # event occurred against dir
+        'IN_ONESHOT'       : 0x80000000,  # only send event once
+        },
+                        }
+
+    def maskname(mask):
+        """
+        Returns the event name associated to mask. IN_ISDIR is appended to
+        the result when appropriate. Note: only one event is returned, because
+        only one event can be raised at a given time.
+
+        @param mask: mask.
+        @type mask: int
+        @return: event name.
+        @rtype: str
+        """
+        ms = mask
+        name = '%s'
+        if mask & IN_ISDIR:
+            ms = mask - IN_ISDIR
+            name = '%s|IN_ISDIR'
+        return name % EventsCodes.ALL_VALUES[ms]
+
+    maskname = staticmethod(maskname)
+
+
+# So let's now turn the configuration into code
+EventsCodes.ALL_FLAGS = {}
+EventsCodes.ALL_VALUES = {}
+for flagc, valc in EventsCodes.FLAG_COLLECTIONS.items():
+    # Make the collections' members directly accessible through the
+    # class dictionary
+    setattr(EventsCodes, flagc, valc)
+
+    # Collect all the flags under a common umbrella
+    EventsCodes.ALL_FLAGS.update(valc)
+
+    # Make the individual masks accessible as 'constants' at globals() scope
+    # and masknames accessible by values.
+    for name, val in valc.items():
+        globals()[name] = val
+        EventsCodes.ALL_VALUES[val] = name
+
+
+# all 'normal' events
+ALL_EVENTS = reduce(lambda x, y: x | y, EventsCodes.OP_FLAGS.values())
+EventsCodes.ALL_FLAGS['ALL_EVENTS'] = ALL_EVENTS
+EventsCodes.ALL_VALUES[ALL_EVENTS] = 'ALL_EVENTS'
+
+
+class _Event:
+    """
+    Event structure, represent events raised by the system. This
+    is the base class and should be subclassed.
+
+    """
+    def __init__(self, dict_):
+        """
+        Attach attributes (contained in dict_) to self.
+
+        @param dict_: Set of attributes.
+        @type dict_: dictionary
+        """
+        for tpl in dict_.items():
+            setattr(self, *tpl)
+
+    def __repr__(self):
+        """
+        @return: Generic event string representation.
+        @rtype: str
+        """
+        s = ''
+        for attr, value in sorted(self.__dict__.items(), key=lambda x: x[0]):
+            if attr.startswith('_'):
+                continue
+            if attr == 'mask':
+                value = hex(getattr(self, attr))
+            elif isinstance(value, str) and not value:
+                value = "''"
+            s += ' %s%s%s' % (output_format.field_name(attr),
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

It appears that you are concatenating a string inside a loop. Concatenating immutable sequences will always results in a new object. This means that building up a sequence by repeated concatenation will have a quadratic runtime cost in the total sequence length. For efficiency and to get a linear runtime cost, consider appending values to a list and concatenating that list using [str.join()](https://docs.python.org/3/library/stdtypes.html#str.join) or [bytes.join()](https://docs.python.org/3/library/stdtypes.html#bytes.join) method.

[Learn more](https://docs.python.org/3/faq/programming.html#what-is-the-most-efficient-way-to-concatenate-many-strings-together)